File: src/emulator/mmu.nim
import bitops
import utils

const RAMBANKS = [0, 1, 2, 4, 16, 8]

var
    bootRom*: string
    romData*: string
    extRam*: seq[uint8]
    wram*: array[0x2000, uint8]
    hram*: array[0x80, uint8]

    romBankCount*: uint8
    ramSizeCode*: uint8
    extRamFlag*: bool

proc buildExternalMemory*(): void =
    extRam = newSeq[uint8](RAMBANKS[ramSizeCode] * 0x2000)

type
    MBC1* = ref object
        romBank*: uint8
        ramBank*: uint8
        modeFlag*: bool

proc read*(m: MBC1, address: uint16): uint8 =
    if address.isboundto(0x0, 0x3FFF):
        result = romData[address].uint8

    elif address.isboundto(0x4000, 0x7FFF):
        var highBankNum = m.romBank and (romBankCount - 1)
        let addrs = 0x4000 * highBankNum + (address - 0x4000)
        result = romData[addrs].uint8

    elif address.isboundto(0xA000, 0xBFFF):
        if extRamFlag:
            if (ramSizeCode > 0x02) and m.modeFlag:
                result = extRam[0x2000 * m.ramBank + (address - 0xA000)]
            else:
                result = extRam[address - 0xA000]
        else:
            result = 0xFF

proc intercept*(m: MBC1, address: uint16, data: uint8): void =
    if address.isboundto(0, 0x1FFF):
        extRamFlag = data.bitsliced(0..3) == 0xA

    elif address.isboundto(0x2000, 0x3FFF):
        let maskedData = (data and 0x1F)
        m.romBank = maskedData mod romBankCount
        if m.romBank == 0 and (maskedData < romBankCount):
            m.romBank = 1

    elif address.isboundto(0x4000, 0x5FFF):
        m.ramBank = data.bitsliced(0..1)

    elif address.isboundto(0x6000, 0x7FFF):
        m.modeFlag = data.testBit(0)
    
    elif address.isboundto(0xA000, 0xBFFF):
        if extRamFlag:
            if (ramSizeCode > 0x02) and m.modeFlag:
                extRam[0x2000 * m.ramBank + (address - 0xA000)] = data
            else:
                extRam[address - 0xA000] = data

File: src/mmu.c
#include "mmu.h"

#include <stdbool.h>
#include <stdint.h>
#include <stdlib.h>

#include "utils.h"

const int RAMBANKS[6] = {0, 1, 2, 4, 16, 8};

uint8_t wram[0x2000];
uint8_t hram[0x80];

uint8_t *bootRom;
uint8_t *romData;
uint8_t *extRam;
uint8_t ramSizeCode;
uint8_t romBankCount;
bool extRamFlag;

bool buildExternalMemory() {
    int totalSize = RAMBANKS[ramSizeCode] * 0x2000;
    extRam = (uint8_t *)malloc(totalSize * sizeof(uint8_t));
    return extRam != NULL;
}

uint8_t readRom(MBC1 *mapper, uint16_t address) {
    if (address <= 0x3FFF) {
        return romData[address];
    }

    else if (BOUND(address, 0x4000, 0x7FFF)) {
        uint16_t highBankNum = mapper->romBank & (romBankCount - 1);
        uint32_t addr = 0x4000 * highBankNum + (address - 0x4000);
        return romData[addr];
    }

    else if (BOUND(address, 0xA000, 0xBFFF)) {
        if (extRamFlag) {
            if ((ramSizeCode > 0x02) && mapper->modeFlag) {
                return extRam[0x2000 * mapper->ramBank + (address - 0xA000)];
            } else {
                return extRam[(address - 0xA000)];
            }
        } else {
            return 0xFF;
        }
    } else {
        return 0xFF;
    }
}

void intercept(MBC1 *mapper, uint16_t address, uint8_t data) {
    if (address <= 0x1FFF) {
        extRamFlag = BEXTR(data, 0, 3) == 0xA;
    } else if (BOUND(address, 0x2000, 0x3FFF)) {
        uint8_t maskedData = data & 0x1F;
        mapper->romBank = maskedData % romBankCount;
        if ((mapper->romBank == 0) && (maskedData < romBankCount)) {
            mapper->romBank = 1;
        }
    } else if (BOUND(address, 0x4000, 0x5FFF)) {
        mapper->ramBank = BEXTR(data, 0, 1);
    } else if (BOUND(address, 0x6000, 0x7FFF)) {
        mapper->modeFlag = BT(data, 0);
    } else if (BOUND(address, 0xA000, 0xBFFF)) {
        if (extRamFlag) {
            if ((ramSizeCode > 0x02) && mapper->modeFlag) {
                extRam[0x2000 * mapper->ramBank + (address - 0xA000)] = data;
            } else {
                extRam[(address - 0xA000)] = data;
            }
        }
    }
}
